docs/proposals-accepted/202206-block-series-deletion-compactor.md

```diff
@@ -0,0 +1,210 @@
+---
+type: proposal
+title: Block Series Deletion with Compactor
+status: accepted
+owner: yeya24
+menu: proposals-accepted
+---
+
+* **Owners:**
+    * @yeya24
+
+## Summary
+
+This design document continues the feature of series deletion for object storage discussed in [proposal](https://thanos.io/tip/proposals-accepted/202012-deletions-object-storage.md/). This proposal is based on the previous work and focuses more on the actual implementation plan and the unaddressed questions from the original proposal like performing the actual deletion in the object storage using Thanos compactor.
+
+## Goals
+
+* Make it more clear about the main workflow of the series deletion process.
+* Perform the actual series deletion using compactor in a scalable way.
+
+## Non-Goals
+
+* Allow setting different retentions per series. This feature can be built on top of this work ideally but it is out of scope of this proposal.
+
+## How
+
+The main workflow of the series deletion will look like this:
+
+1. User create a tombstone file via a CLI tool and the tombstone file is uploaded to the object storage.
+2. Store Gateway syncs tombstone files periodically and makes the block series based on the tombstones.
+3. Tombstone files have a pending period, which can be either specified by each tombstone or configured globally. Before the pending period is over, the tombstone file will remain in the object storage and user can remove a tombstone file using the CLI tool. Once the pending period is over, the compactor starts processing the tombstones and performs deletion.
+4. For each tombstone file, once all the matched blocks are processed, compactor will delete the tombstone file from the object storage.
+
+### Tombstone File
+
+The design of the tombstone file mainly follows the previous proposal. Tombstone file is immutable. It is using a custom JSON format, single file per request and global, which means it applies to all the blocks. For the detailed considerations and tradeoffs, please refer to the original proposal.
+
+The file format is as follows:
+``` bash
+cat 01G61GC0Q957TVZ2FC8S6Y2ZK7.json
+{
+        "ulid": "01G61GC0Q957TVZ2FC8S6Y2ZK7",
+        "matchers": [
+                "__name__=\"k8s_app_metric39\"",
+                "cluster=\"us\"",
+                "__block_id=\"01G61D0Q4BHTRC7HW08YRD0KGS\""
```

Comment:
Yes. This is something I added. But block ID is an optional field, which means users can choose to not add it and then the matchers can match all blocks. If they choose to add it then they can delete the specific block they want to. (Even multiple blocks using =~)

Reply:
Can we use something like the prometheus relabel configs, or is that too complicated for now?

Reply:
Hi Filip, if I understand this correctly, the tombstone file is global in the bucket and each tombstone file can match all the blocks (if no external label marcher and block id marchers are specified).
So if we are talking about sharding, store gateway and compactor need to gather all tombstones to match their blocks. Their blocks are sharded already, but tombstones are not.

That might be an issue when sharded compactors are running. Because each compactor only has view for its own blocks, we don't know whether a tombstone file has been applied to all blocks or not. So some kind of compactor coordinations are needed. WDYT about this issue?

Reply:
If we want to shard on tombstones then we might need per block tombstone file…

Reply:
What I had in mind was to use relabel configs instead of explicit matchers to keep or drop individual series. But I guess that will increase the scope of the proposal to allow complete series rewriting in compactor, not just deletion. So something to keep in mind for the future, but too complicated for implementing in the first version.

Reply:
Oh I see what you meant. Sorry I was talking about different things... 
Similar idea has been mentioned in https://github.com/thanos-io/thanos/issues/4941. We supported this in `bucket rewrite tool` and I proposed similar things to Prometheus https://github.com/prometheus/prometheus/pull/9413 before.

But indeed this has a different scope. Maybe it is better to think about it when we are working on the per-series retention period feature on compactor https://github.com/thanos-io/thanos/issues/903.

Reply:
Yes, we discussed this - we could add new field for tombstones to support rewrites at some point.

I like block ID thingy @yeya24  - we could add that. Just let's fix the sentence that it applies to all blocks, unless block ID is specified

Reply:
> ust let's fix the sentence that it applies to all blocks, unless block ID is specified

The matchers support matching external labels, too for multi-tenancy purposes. I think we can still call it global and it applies to all blocks. Just we can tell that the matchers won't match some blocks so we just filter them out. Like how we do requests filtering on store gateway.